File: Data Structures & Algorithms/valid-palindrome-ii/submission-3.py
class Solution:
    def validPalindrome(self, s: str) -> bool:

        if self.isPalindrome(s) or len(s) <= 2: return True

        l=0
        r=len(s)-1

        while l<r:
            if s[l] != s[r]:
                return self.isPalindrome(s[l:r]) or self.isPalindrome(s[l+1:r+1])
            l += 1
            r -= 1
        
        return True

    def isPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s)-1

        while l < r:
            if s[l] != s[r]:
                return False
            l += 1
            r -=1
        return True

File: Data Structures & Algorithms/is-palindrome/submission-3.java
class Solution {
    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length()-1;

        while (left < right) {
            if (!this.isAlpha(s.charAt(left))) {
                left++;
            } else if (!this.isAlpha(s.charAt(right))) {
                right--;
            } else {
                if (s.toLowerCase().charAt(left) != s.toLowerCase().charAt(right)) return false;
                left++;
                right--;
            }
        }

        return true;
    }

    public boolean isAlpha(char c){
        return (c <= 'z' && c >= 'a') 
            || (c <= 'Z' && c >= 'A')
            || Character.isDigit(c);
    }
}
